genic/main.c

```diff
@@ -88,15 +89,35 @@ int main(int argc, char **argv)
   }
 
   /*First compute and write CDM*/
+  double mass[6] = {0};
+  /*Can neglect neutrinos since this only matters for the glass force.*/
+  compute_mass(mass, TotNumPart, TotNumPartGas, 0, 0);
   /*Not used*/
   int size[3], offset[3];
   int NumPart = get_size_offset(size, offset, All2.Ngrid);
-  struct ic_part_data * ICP = (struct ic_part_data *) mymalloc("PartTable", NumPart*sizeof(struct ic_part_data));
+  int NumPartGas = get_size_offset(size, offset, All2.NgridGas);
+  /*Space for both CDM and baryons*/
+  struct ic_part_data * ICP = (struct ic_part_data *) mymalloc("PartTable", (NumPart + All2.ProduceGas * NumPartGas)*sizeof(struct ic_part_data));
 
+  /* If we have incoherent glass files, we need to store both the particle tables
+   * to ensure that there are no close particle pairs*/
+  /*Make the table for the CDM*/
   if(!All2.MakeGlassCDM) {
-      setup_grid(All2.ProduceGas * shift_dm, All2.Ngrid, NumPart, ICP);
+      setup_grid(All2.ProduceGas * shift_dm, All2.Ngrid, mass[1], NumPart, ICP);
   } else {
-      setup_glass(0, All2.Ngrid, GLASS_SEED_HASH(All2.Seed), NumPart, ICP);
+      setup_glass(0, All2.Ngrid, GLASS_SEED_HASH(All2.Seed), mass[1], NumPart, ICP);
+  }
+
+  /*Make the table for the baryons if we need, using the second half of the memory.*/
+  if(All2.ProduceGas) {
+    if(!All2.MakeGlassGas) {
+        setup_grid(shift_gas, All2.NgridGas, mass[0], NumPart, ICP+NumPart);
+    } else {
+        setup_glass(0, All2.NgridGas, GLASS_SEED_HASH(All2.Seed + 1), mass[0], NumPart, ICP+NumPart);
+    }
+    /*Do a single glass evolution timestep to avoid close pairs*/
+    if(All2.MakeGlassGas || All2.MakeGlassCDM)
+        glass_evolve(14, 0xDEADBEEF, ICP, NumPart+NumPartGas);
```

Comment:
Be aware, this says 14 steps, not a single step. 

Is it feasible to split ICP to ICPCDM, ICPGas and ICPNu? It may look leaner than passing around both the base pointer and the offset.

Reply:
I will fix the comment (it is a relic). 

I prefer not to split ICP, I think, to make it explicit that it is the same allocation and thus one can run the glass on the combined particle list.